#Commands to run RNA-Seq analysis using R ballgown package
#Numbers after the hashtag (#) sysmbol corresponds to sections in manual

## 1. Call the ballgown library as follows:
library(ballgown);library(genefilter);library(dplyr);library(RColorBrewer); library(ggfortify);library(devtools);library(gplots);

## 2. Assign the path to directory containing ballgown outputs from StringTie to a variable
data_directory = ('~/Desktop/hisat/ballgown')

## 3. Reading in the Outputs from StringTie
bg = ballgown(dataDir=data_directory, samplePattern='r', meas='all')

## 4. Print the structure of our files
bg

## 5. Create metadata information for the files. Ballgown doesn't perform DE analysis
## without replicates. For convenience, we're combining Sp_ds & Sp_hs as replicates
## and Sp_plat & Sp_log as replicates.

pData(bg) = data.frame(id=sampleNames(bg), group=factor(c(1,1,2,2)))

## 6. What type of information was genereated by StringTie
names(structure(bg))

## 7. Looking individually at each of the transcript assembly types generated by StringTie
structure(bg)$exon
structure(bg)$intron
structure(bg)$trans

## 8. Isolate FPKM values for each transcript
transcript_fpkm = texpr(bg, 'FPKM')

## 9. Isolate coverage information for each transcript
transcript_cov = texpr(bg, 'cov')

## Take a look at this webpage:
#http://bioconductor.org/packages/release/bioc/vignettes/ballgown/inst/doc/ballgown.html

## 10. Isolate all available stats for each transcript
whole_tx_table = texpr(bg, 'all')

## 11. Isolate multi-map-corrected average per-base read coverage for each exon
exon_mcov = eexpr(bg, 'mcov')

## 12. Isolate read count for each intron
junction_rcount = iexpr(bg)

## 13. Isolate all available stats for each intron
whole_intron_table = iexpr(bg, 'all')

## 14. Isolate gene-level expression
gene_expression = gexpr(bg)

# 15. Transform FPKM to log2 scale for visualization
transcript_fpkm_log2 = log2(transcript_fpkm+1)

#Boxplot
par(mar=c(10,4,4,2))
boxplot(transcript_fpkm_log2,col=as.numeric(pData(bg)$group),las=2,ylab='log2(FPKM+1)')
dev.off()

# 16. heatmap
cols = colorpanel(99,"blue","black","yellow")
heatmap.2(transcript_fpkm_log2,density.info = "none",trace ="none",margins = c(10,15), cexCol = 1.2,keysize = 1,col = cols,scale = "none")

# 17. PCA
dev.off()
pca = prcomp(t(transcript_fpkm_log2))
autoplot(pca,data=pData(bg),colour="group",label=T,label.label="id")

# 18. Extract indexes from ballgown object
names(indexes(bg))

# 19. mapping exons to transcripts
head(indexes(bg)$e2t)

# 20. mapping introns to transcripts
head(indexes(bg)$i2t)

# 21. mapping transcripts to genes
head(indexes(bg)$t2g)

# 22. mapping exons to transcripts and transcripts to gene
exon_transcript_table = indexes(bg)$e2t
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)

# 23. statistical test to identify differential expression by transcript & gene
results_transcripts = stattest(bg, feature='transcript', meas='FPKM',covariate='group',getFC=TRUE)
head(results_transcripts)
write.table(results_transcripts,file="ballgown_transcript_results.txt",sep="\t",quote=F,row.names = F)

results_genes = stattest(bg, feature='gene', meas='FPKM',covariate='group',getFC=TRUE)
head(results_genes)
write.table(results_genes,file="ballgown_gene_results.txt",sep="\t",quote=F,row.names = F)

# 24. Adding gene ids to trancript results
results_transcripts =data.frame(geneIDs=ballgown::geneIDs(bg), results_transcripts)
head(results_transcripts)

results_transcripts = arrange(results_transcripts,pval)
results_genes = arrange(results_genes,pval)

## 25. Plot a map of this transcript across all the samples and color by transcript
head(results_transcripts)
plotTranscripts(gene='MSTRG.135', gown=bg,samples = sampleNames(bg), meas='FPKM', colorby='transcript',main='transcripts from gene MSTRG.135:  FPKM')
dev.off()

# 26. boxplot showing log2 FPKM of most differentially expressed transcript
idx = which(geneIDs(bg)=="MSTRG.135")
idx
plot(transcript_fpkm_log2[idx,] ~ pData(bg)$group, border=c(1,2),main=paste(ballgown::geneIDs(bg)[idx]),pch=19, xlab="Group",ylab='log2(FPKM+1)')
points(transcript_fpkm_log2[idx,] ~ jitter(as.numeric(pData(bg)$group)),col=as.numeric(pData(bg)$group),pch=20,cex=1.5)

## Free to use/modify/resuse for personal academic purposes
##Â© 2020 American Academy for Biomedical Informatics, Confidential and Proprietary 
